{"cells":[{"cell_type":"code","source":["import numpy as np\n#import findspark\n#findspark.init()\nimport pyspark as ps\nimport re\nimport string\nfrom pyspark import SparkContext\nfrom pyspark.ml.feature import Tokenizer, HashingTF, IDF\nfrom pyspark.ml.feature import CountVectorizer\nfrom pyspark.sql import SQLContext\nfrom pyspark.ml.feature import StopWordsRemover\nfrom pyspark.ml.feature import Normalizer\nfrom sklearn.metrics.pairwise import cosine_similarity"],"metadata":{"collapsed":true},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":1},{"cell_type":"code","source":["# File location and type\nfile_location = \"/FileStore/tables/data_old.csv\"\n#file_location = \"/FileStore/tables/data.csv\"\nfile_type = \"csv\"\n\n# CSV options\ninfer_schema = \"false\"\nfirst_row_is_header = \"false\"\ndelimiter = \"\\t\"\n\n# The applied options are for CSV files. For other file types, these will be ignored.\ndf = spark.read.format(file_type) \\\n  .option(\"inferSchema\", infer_schema) \\\n  .option(\"header\", first_row_is_header) \\\n  .option(\"sep\", delimiter) \\\n  .load(file_location)\n\ndisplay(df.head(3))"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .table-result-container {\n    max-height: 300px;\n    overflow: auto;\n  }\n  table, th, td {\n    border: 1px solid black;\n    border-collapse: collapse;\n  }\n  th, td {\n    padding: 5px;\n  }\n  th {\n    text-align: left;\n  }\n</style><div class='table-result-container'><table class='table-result'><thead style='background-color: white'><tr><th>_c0</th><th>_c1</th><th>_c2</th><th>_c3</th><th>_c4</th><th>_c5</th><th>_c6</th><th>_c7</th></tr></thead><tbody><tr><td>id</td><td>title</td><td>authors</td><td>year</td><td>journal</td><td>abstract</td><td>tags</td><td>citations</td></tr><tr><td>27999001</td><td>Glucose Metabolism After Gastric Banding and Gastric Bypass in Individuals With Type 2 Diabetes: Weight Loss Effect.</td><td>['Holter', 'Dutia', 'Stano', 'Prigeon', 'Homel', 'McGinty', 'Belsley', 'Ren', 'Rosen', 'Laferrère']</td><td>2017</td><td>Diabetes care</td><td>The superior effect of Roux-en-Y gastric bypass (RYGB) on glucose control compared with laparoscopic adjustable gastric banding (LAGB) is confounded by the greater weight loss after RYGB. We therefore examined the effect of these two surgeries on metabolic parameters matched on small and large amounts of weight loss.</td><td>['Adult', 'Bariatric Surgery', 'Diabetes Mellitus, Type 2', 'Female', 'Gastric Bypass', 'Glucagon-Like Peptide 1', 'Glucose', 'Humans', 'Incretins', 'Insulin Resistance', 'Longitudinal Studies', 'Male', 'Middle Aged', 'Obesity', 'Postoperative Period', 'Prospective Studies', 'Sweetening Agents', 'Weight Loss']</td><td>['26826918', '26132586', '22535748', '24679060', '18650633', '20676394', '23388352', '24114113', '24296713', '18070760', '26186884', '26807004', '20029383', '26681719', '21107106', '23649520', '23652711', '23439632', '21078684', '7677463', '23187122', '12086947', '22449317', '16600932', '24089513', '27289123', '443421', '1551497', '25628424', '24189773', '23610060', '22359255', '24057293', '21339424', '26786780', '16478824', '20716694', '18430778', '17416796', '3899825']</td></tr><tr><td>27999002</td><td>Metformin Is Associated With Higher Relative Abundance of Mucin-Degrading Akkermansia muciniphila and Several Short-Chain Fatty Acid-Producing Microbiota in the Gut.</td><td>['de la Cuesta-Zuluaga', 'Mueller', 'Corrales-Agudelo', 'Velásquez-Mejía', 'Carmona', 'Abad', 'Escobar']</td><td>2017</td><td>Diabetes care</td><td>Recent studies suggest the beneficial effects of metformin on glucose metabolism may be microbially mediated. We examined the association of type 2 diabetes, metformin, and gut microbiota in community-dwelling Colombian adults. On the basis of previous research, we hypothesized that metformin is associated with higher levels of short-chain fatty acid (SCFA)-producing and mucin-degrading microbiota.</td><td>['Adolescent', 'Adult', 'Case-Control Studies', 'Colombia', 'Diabetes Mellitus, Type 2', 'Fatty Acids, Volatile', 'Feces', 'Female', 'Gastrointestinal Microbiome', 'Humans', 'Hypoglycemic Agents', 'Male', 'Metformin', 'Middle Aged', 'Mucins', 'RNA, Ribosomal, 16S', 'Verrucomicrobia']</td><td>['27910881']</td></tr></tbody></table></div>"]}}],"execution_count":2},{"cell_type":"code","source":["#import findspark\n#findspark.init(\"/usr/local/Cellar/apache-spark/2.3.1/bin\")"],"metadata":{"collapsed":true},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":3},{"cell_type":"code","source":["def remove_punctuation(text):\n    \"\"\" This method removes the punctuation like commas and quotes from the text (string). \n    We also want to keep contractions together. The method also make the words lower cased.\n    It returns a list or words in the text\n        Args:\n            text (string): the text we want to clean\n        Return:\n            A list with cleaned words\n    \"\"\"\n    # split into words by white space\n    words = text.split()\n    words_lower = [w.lower() for w in words]\n    \n    # Remove punctuation from each word\n    table = str.maketrans('', '', string.punctuation)\n    stripped = [w.translate(table) for w in words_lower]\n    return ' '.join(stripped)"],"metadata":{"collapsed":true},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":4},{"cell_type":"code","source":["# get the similarities for each pair of papers\ndef get_similarities(paper_rdd, paper_tfidf):\n    \"\"\" Function that returns the array of similarities between each two papers\n        Args:\n            paper_rdd (rdd): idd of all papers abstract and titles\n            paper_tfidf (pyspark.sql.dataframe.DataFrame): tf-idf vectors for a given paper\n        Return:\n            similarity_array:  array of cosine similarities for each pair of papers\n    \"\"\"\n    \n    print(\"... Computing L2 norm ...\")\n    labels = paper_rdd.map(lambda x: x[0])\n    features = paper_tfidf\n\n    normalizer = Normalizer(inputCol=\"features\", outputCol=\"normFeatures\")\n    data = labels.zip(normalizer.transform(features).rdd.map(lambda r: r.normFeatures))\n    \n    #Using a Cartesian product and the function dot on numpy arrays:\n    similarity_array = data.cartesian(data)\\\n    .map(lambda l: ((l[0][0], l[1][0]), l[0][1].dot(l[1][1])))\\\n    .sortByKey()\n    \n    return similarity_array"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":5},{"cell_type":"code","source":["# get the n top similar papers for given paper info\ndef get_neighbors(paper_PMID, similarity_array, n):\n    \"\"\" Function that returns the 50 most similar papers for given paper\n        Args:\n            paper_PMID (int): PMID of the paper we want to find similar papers to\n            similarity_array (array): cosine similarity array for all papers\n            n (int): number of similar papers we are looking for, for a specified paper\n        Return:\n            list: the list of papers relevant to the given paper based on cosine similarity\n    \"\"\"\n    candidates = similarity_array.filter(lambda x: x[0][0]==paper_PMID).sortBy(lambda a: -a[1])\n    neighbors = candidates.map(lambda x: x[0][1])\n\n    return neighbors.take(n)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":6},{"cell_type":"code","source":["def paper_recommender_title_abstract(paper_in_PMID, text_rdd, n):\n    \"\"\"The main function that gets a PMID for a paper, and recommends n similar papers, using the metadata (title and abstract) of all papers in text_rdd\"\"\"\n    #sqlCtx = SQLContext(sc)\n    \n    text_rdd = text_rdd.zipWithIndex()\n    \n    # Separate the header from the text file\n    header = text_rdd.first()\n    \n    # Remove the header from the text_rdd\n    text_rdd_n = text_rdd.filter(lambda x: x != header).map(lambda y: (y[1], y[0].split('\\t')))\n\n    # Extract title and abstract for each paper and remove punctuations from them\n    print(\"... Cleaning the Title & Abstract ...\")    \n    paper_abs_title_rdd = text_rdd_n.map(lambda x: (x[1][0], remove_punctuation(x[1][1] + \" \" +x[1][5])))       \n    sentenceData_df = sqlContext.createDataFrame(paper_abs_title_rdd, [\"label\", \"sentence\"])\n    \n    print(\"... Tokenizing and StopWords removing ...\")\n    tokenizer = Tokenizer(inputCol=\"sentence\", outputCol=\"words\")\n    wordsData = tokenizer.transform(sentenceData_df)\n    remover = StopWordsRemover(inputCol=\"words\", outputCol=\"filtered\")\n    wordsData = remover.transform(wordsData)#.show(truncate=False)\n    \n    print(\"... Computing TF-IDF ...\")\n    hashingTF = HashingTF(inputCol=\"filtered\", outputCol=\"rawFeatures\")\n    tf = hashingTF.transform(wordsData).cache()\n    \n    idfModel = IDF(inputCol=\"rawFeatures\", outputCol=\"features\")\n    idf = idfModel.fit(tf)\n    tfidf = idf.transform(tf)\n\n    print(\"... Calculating similarity array ...\")\n    data_similarity = get_similarities(paper_abs_title_rdd, tfidf)\n    \n    print(\"... Finding n similar papers ...\")\n    neighbors = get_neighbors(paper_in_PMID, data_similarity, n)\n    \n    print(\"Neighbors based on title and abstract =\", neighbors)\n    \n    print(\"Main paper title: \", text_rdd_n.filter(lambda x: x[1][0]==paper_in_PMID).map(lambda y: y[1][1]).collect())\n    \n    for i in neighbors:\n      print(\"Similar paper title for PMID\", i, \":\", text_rdd_n.filter(lambda x: x[1][0]==i).map(lambda y: y[1][1]).collect())\n\n    return()"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":7},{"cell_type":"code","source":["def paper_recommender_tags(paper_in_PMID, text_rdd, n):\n    \"\"\"The main function that gets a PMID for a paper, and recommends n similar papers, using the tags of all papers in text_rdd\"\"\"\n    #sqlCtx = SQLContext(sc)\n    \n    text_rdd = text_rdd.zipWithIndex()\n    \n    # Separate the header from the text file\n    header = text_rdd.first()\n    \n    # Remove the header from the text_rdd\n    text_rdd_n = text_rdd.filter(lambda x: x != header).map(lambda y: (y[1], y[0].split('\\t')))\n\n    # Extract the tags for each paper and remove punctuations from them\n    print(\"... Cleaning the tags ...\")\n    \n    paper_tags_rdd = text_rdd_n.map(lambda x: (x[1][0],x[1][6].lower()))\n    \n    sentenceData_df = sqlContext.createDataFrame(paper_tags_rdd, [\"label\", \"sentence\"])\n    \n    print(\"... Tokenizing and StopWords removing ...\")\n    \n    tokenizer = Tokenizer(inputCol=\"sentence\", outputCol=\"words\")\n    wordsData = tokenizer.transform(sentenceData_df)\n    remover = StopWordsRemover(inputCol=\"words\", outputCol=\"filtered\")\n    wordsData = remover.transform(wordsData)\n    \n    print(\"... Computing TF-IDF ...\")\n    \n    hashingTF = HashingTF(inputCol=\"filtered\", outputCol=\"rawFeatures\")\n    tf = hashingTF.transform(wordsData).cache()\n    \n    idfModel = IDF(inputCol=\"rawFeatures\", outputCol=\"features\")\n    idf = idfModel.fit(tf)\n    tfidf = idf.transform(tf)\n    \n    print(\"... Calculating similarity array ...\")\n    data_similarity = get_similarities(paper_tags_rdd, tfidf)\n    \n    print(\"... Finding n similar papers ...\")\n    neighbors = get_neighbors(paper_in_PMID, data_similarity, n)\n    \n    print(\"Neighbors based on tags =\", neighbors)\n    \n    print(\"Main paper tags: \", text_rdd_n.filter(lambda x: x[1][0]==paper_in_PMID).map(lambda y: y[1][6]).collect())\n    \n    for i in neighbors:\n      print(\"Similar paper tags for PMID\", i, \":\", text_rdd_n.filter(lambda x: x[1][0]==i).map(lambda y: y[1][6]).collect())\n\n    return()"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":8},{"cell_type":"code","source":["if __name__ == '__main__':\n    #sc.stop()\n    #sc = ps.SparkContext('local[1]')\n    text_rdd = sc.textFile(file_location)\n\n    print(\"... Finding similar papers based on title and abstract ...\")\n    #paper_recommender_title_abstract('26990000', text_rdd, 10)    \n    paper_recommender_title_abstract('27999001', text_rdd, 10) \n    \n    print(\"--------------------------------------------------\")\n    print(\"... Finding similar papers based on tags ...\")\n    #paper_recommender_tags('26990000', text_rdd, 10) \n    paper_recommender_tags('27999001', text_rdd, 10) "],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">... Finding similar papers based on title and abstract ...\n... Cleaning the Title &amp; Abstract ...\n... Tokenizing and StopWords removing ...\n... Computing TF-IDF ...\n... Calculating similarity array ...\n... Computing L2 norm ...\n... Finding n similar papers ...\nNeighbors based on title and abstract = [&apos;27999001&apos;, &apos;27999914&apos;, &apos;27999962&apos;, &apos;27999678&apos;, &apos;27999812&apos;, &apos;27999842&apos;, &apos;27999039&apos;, &apos;27999105&apos;, &apos;27999104&apos;, &apos;27999311&apos;]\nMain paper title:  [&apos;Glucose Metabolism After Gastric Banding and Gastric Bypass in Individuals With Type 2 Diabetes: Weight Loss Effect.&apos;]\nSimilar paper title for PMID 27999001 : [&apos;Glucose Metabolism After Gastric Banding and Gastric Bypass in Individuals With Type 2 Diabetes: Weight Loss Effect.&apos;]\nSimilar paper title for PMID 27999914 : [&apos;Laparoscopic sentinel node navigation surgery for early gastric cancer: a prospective multicenter trial.&apos;]\nSimilar paper title for PMID 27999962 : [&apos;Can lymphovascular invasion be predicted by preoperative multiphasic dynamic CT in patients with advanced gastric cancer?&apos;]\nSimilar paper title for PMID 27999678 : [&apos;POU4F3 Gene Causes Autosomal Dominant Hearing Loss.&apos;]\nSimilar paper title for PMID 27999812 : [&apos;Glucose &gt;200\\u2009mg/dL during Continuous Glucose Monitoring Identifies Adult Patients at Risk for Development of Cystic Fibrosis Related Diabetes.&apos;]\nSimilar paper title for PMID 27999842 : [&apos;Effect of a High-Protein Energy-Restricted Diet Combined with Resistance Training on Metabolic Profile in Older Individuals with Metabolic Impairments.&apos;]\nSimilar paper title for PMID 27999039 : [&apos;Population pharmacokinetics of cefazolin before, during and after cardiopulmonary bypass to optimize dosing regimens for children undergoing cardiac surgery.&apos;]\nSimilar paper title for PMID 27999105 : [&apos;Insulin Regulates Astrocytic Glucose Handling Through Cooperation With IGF-I.&apos;]\nSimilar paper title for PMID 27999104 : [&apos;Heparanase Overexpression Induces Glucagon Resistance and Protects Animals From Chemically Induced Diabetes.&apos;]\nSimilar paper title for PMID 27999311 : [&apos;Exercise and Beta-Glucan Consumption (Saccharomyces cerevisiae) Improve the Metabolic Profile and Reduce the Atherogenic Index in Type 2 Diabetic Rats (HFD/STZ).&apos;]\n--------------------------------------------------\n... Finding similar papers based on tags ...\n... Cleaning the tags ...\n... Tokenizing and StopWords removing ...\n... Computing TF-IDF ...\n... Calculating similarity array ...\n... Computing L2 norm ...\n... Finding n similar papers ...\nNeighbors based on tags = [&apos;27999001&apos;, &apos;27999351&apos;, &apos;27999104&apos;, &apos;27999135&apos;, &apos;27999002&apos;, &apos;27999873&apos;, &apos;27999311&apos;, &apos;27999106&apos;, &apos;27999865&apos;, &apos;27999950&apos;]\nMain paper tags:  [&quot;[&apos;Adult&apos;, &apos;Bariatric Surgery&apos;, &apos;Diabetes Mellitus, Type 2&apos;, &apos;Female&apos;, &apos;Gastric Bypass&apos;, &apos;Glucagon-Like Peptide 1&apos;, &apos;Glucose&apos;, &apos;Humans&apos;, &apos;Incretins&apos;, &apos;Insulin Resistance&apos;, &apos;Longitudinal Studies&apos;, &apos;Male&apos;, &apos;Middle Aged&apos;, &apos;Obesity&apos;, &apos;Postoperative Period&apos;, &apos;Prospective Studies&apos;, &apos;Sweetening Agents&apos;, &apos;Weight Loss&apos;]&quot;]\nSimilar paper tags for PMID 27999001 : [&quot;[&apos;Adult&apos;, &apos;Bariatric Surgery&apos;, &apos;Diabetes Mellitus, Type 2&apos;, &apos;Female&apos;, &apos;Gastric Bypass&apos;, &apos;Glucagon-Like Peptide 1&apos;, &apos;Glucose&apos;, &apos;Humans&apos;, &apos;Incretins&apos;, &apos;Insulin Resistance&apos;, &apos;Longitudinal Studies&apos;, &apos;Male&apos;, &apos;Middle Aged&apos;, &apos;Obesity&apos;, &apos;Postoperative Period&apos;, &apos;Prospective Studies&apos;, &apos;Sweetening Agents&apos;, &apos;Weight Loss&apos;]&quot;]\nSimilar paper tags for PMID 27999351 : [&quot;[&apos;Coronary Artery Disease&apos;, &apos;Diabetes Complications&apos;, &apos;Diabetes Mellitus, Type 1&apos;, &apos;Diabetes Mellitus, Type 2&apos;, &apos;Diabetic Cardiomyopathies&apos;, &apos;Glycated Hemoglobin A&apos;, &apos;Heart Failure&apos;, &apos;Humans&apos;]&quot;]\nSimilar paper tags for PMID 27999104 : [&quot;[&apos;Animals&apos;, &apos;Diabetes Mellitus, Experimental&apos;, &apos;Fibroblast Growth Factors&apos;, &apos;Glucagon&apos;, &apos;Glucagon-Like Peptide 1&apos;, &apos;Glucuronidase&apos;, &apos;Hyperglycemia&apos;, &apos;Insulin&apos;, &apos;Islets of Langerhans&apos;, &apos;Male&apos;, &apos;Mice&apos;, &apos;Mice, Inbred C57BL&apos;, &apos;Mice, Transgenic&apos;, &apos;Streptozocin&apos;]&quot;]\nSimilar paper tags for PMID 27999135 : [&quot;[&apos;Diabetes Mellitus, Type 2&apos;, &apos;Humans&apos;, &apos;Hypoglycemic Agents&apos;, &apos;Insulin Resistance&apos;, &apos;Prediabetic State&apos;, &apos;Randomized Controlled Trials as Topic&apos;, &apos;Secondary Prevention&apos;, &apos;Stroke&apos;, &apos;Thiazolidinediones&apos;]&quot;]\nSimilar paper tags for PMID 27999002 : [&quot;[&apos;Adolescent&apos;, &apos;Adult&apos;, &apos;Case-Control Studies&apos;, &apos;Colombia&apos;, &apos;Diabetes Mellitus, Type 2&apos;, &apos;Fatty Acids, Volatile&apos;, &apos;Feces&apos;, &apos;Female&apos;, &apos;Gastrointestinal Microbiome&apos;, &apos;Humans&apos;, &apos;Hypoglycemic Agents&apos;, &apos;Male&apos;, &apos;Metformin&apos;, &apos;Middle Aged&apos;, &apos;Mucins&apos;, &apos;RNA, Ribosomal, 16S&apos;, &apos;Verrucomicrobia&apos;]&quot;]\nSimilar paper tags for PMID 27999873 : [&quot;[&apos;Adolescent&apos;, &apos;Adult&apos;, &apos;Area Under Curve&apos;, &apos;Child&apos;, &apos;Diabetes Mellitus, Type 2&apos;, &apos;Dipeptidyl-Peptidase IV Inhibitors&apos;, &apos;Female&apos;, &apos;Half-Life&apos;, &apos;Humans&apos;, &apos;Hypoglycemic Agents&apos;, &apos;Male&apos;, &apos;Piperidines&apos;, &apos;Uracil&apos;, &apos;Young Adult&apos;]&quot;]\nSimilar paper tags for PMID 27999311 : [&quot;[&apos;Animals&apos;, &apos;Atherosclerosis&apos;, &apos;Biomarkers&apos;, &apos;Blood Glucose&apos;, &apos;Combined Modality Therapy&apos;, &apos;Diabetes Mellitus, Experimental&apos;, &apos;Diabetes Mellitus, Type 2&apos;, &apos;Diabetic Angiopathies&apos;, &apos;Diabetic Nephropathies&apos;, &apos;Diet, High-Fat&apos;, &apos;Dietary Fiber&apos;, &apos;Dietary Supplements&apos;, &apos;Exercise Therapy&apos;, &apos;Glycated Hemoglobin A&apos;, &apos;Lipids&apos;, &apos;Male&apos;, &apos;Rats, Wistar&apos;, &apos;Saccharomyces cerevisiae&apos;, &apos;Streptozocin&apos;, &apos;beta-Glucans&apos;]&quot;]\nSimilar paper tags for PMID 27999106 : [&quot;[&apos;Animals&apos;, &apos;CRISPR-Cas Systems&apos;, &apos;Cells, Cultured&apos;, &apos;Diabetes Mellitus, Type 1&apos;, &apos;Female&apos;, &apos;Immunity, Innate&apos;, &apos;Interferon Type I&apos;, &apos;Male&apos;, &apos;Parvovirus&apos;, &apos;Rats&apos;, &apos;Real-Time Polymerase Chain Reaction&apos;, &apos;Receptor, Interferon alpha-beta&apos;, &apos;Reverse Transcriptase Polymerase Chain Reaction&apos;]&quot;]\nSimilar paper tags for PMID 27999865 : [&quot;[&apos;Aged&apos;, &apos;California&apos;, &apos;Cohort Studies&apos;, &apos;Comorbidity&apos;, &apos;Female&apos;, &apos;Florida&apos;, &apos;Humans&apos;, &apos;Kidney Calculi&apos;, &apos;Length of Stay&apos;, &apos;Male&apos;, &apos;Middle Aged&apos;, &apos;Mortality&apos;, &apos;Nephrolithotomy, Percutaneous&apos;, &apos;Paralysis&apos;, &apos;Perioperative Period&apos;, &apos;Postoperative Complications&apos;, &apos;Sepsis&apos;, &apos;Spinal Cord Injuries&apos;]&quot;]\nSimilar paper tags for PMID 27999950 : [&quot;[&apos;Aged&apos;, &apos;Aging&apos;, &apos;China&apos;, &apos;Chronic Disease&apos;, &apos;Cross-Sectional Studies&apos;, &apos;Diabetes Mellitus&apos;, &apos;Dyslipidemias&apos;, &apos;Female&apos;, &apos;Humans&apos;, &apos;Hypertension&apos;, &apos;Male&apos;, &apos;Middle Aged&apos;, &apos;Noncommunicable Diseases&apos;, &apos;Obesity&apos;, &apos;Osteoporosis&apos;, &apos;Prevalence&apos;, &apos;Risk Factors&apos;, &apos;Surveys and Questionnaires&apos;]&quot;]\n</div>"]}}],"execution_count":9}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"mimetype":"text/x-python","name":"python","pygments_lexer":"ipython3","codemirror_mode":{"name":"ipython","version":"3"},"version":"3.6.2","nbconvert_exporter":"python","file_extension":".py"},"name":"recommender","notebookId":2162429554180729},"nbformat":4,"nbformat_minor":0}
